aiomonitor/monitor.py

```diff
@@ -154,85 +232,98 @@ def _filter_cmds(self, *,
                     if altname.startswith(startswith):
                         yield CmdName(altname, name)
 
-    def _command_dispatch(self,
-                          sin: IO[str],
-                          sout: IO[str],
-                          user_input: str) -> None:
-        comm, *args = user_input.split(' ')
-        cmd = self._getcmd(comm)
-        try:
-            cmd(sin, sout, *args)
-        except Exception as e:
-            msg = 'Exception occured during command "{}": {}'
-            msg = msg.format(user_input, repr(e))
-            sout.write(msg + '\n')
-            log.exception(msg)
-
-    def _getcmd(self, comm: str) -> Callable:
+    def map_args(self, cmd: Callable, args: Sequence[str]
+                 ) -> Generator[Any, None, None]:
+        params = inspect.signature(cmd).parameters.values()
+        ia = iter(args)
+        for param in params:
+            if (param.annotation is param.empty or
+                    not callable(param.annotation)):
+                type_ = lambda x: x  # noqa
+            else:
+                type_ = param.annotation
+            try:
+                if str(param).startswith('*'):
+                    yield from (type_(arg) for arg in ia)  # type: ignore
+                else:
+                    yield type_(next(ia))  # type: ignore
```

Comment:
we expect StopIteration only here?

Reply:
We iterate over the functions annotation for its parameters. Within that, we manually iterate over the given arguments (with the `ia = iter(args)` and `next(ia)` construct; I don't use `zip` since those two sequences may have arbitrary length difference). Now, when when we are in the `else` clause of the `if` (meaning we have not yet reached `*args`), we could be in the situation where a further parameter exists, but no argument is given to that parameter. Since we might have `do_*` methods with optional, non-star arguments, we must ignore a `StopIteration` from this call to `next`.
I probably should add this as a comment to the code.